File: test/ModularPipelines.UnitTests/Attributes/DynamicDependencyIntegrationTests.cs
using ModularPipelines.Attributes.Events;
using ModularPipelines.Context;
using ModularPipelines.Exceptions;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Attributes;

[NotInParallel(nameof(DynamicDependencyIntegrationTests))]
public class DynamicDependencyIntegrationTests : TestBase
{
    private static readonly List<string> ExecutionOrder = new();

    public class AddDependencyAttribute : Attribute, IModuleRegistrationEventReceiver
    {
        private readonly Type _dependencyType;

        public AddDependencyAttribute(Type dependencyType)
        {
            _dependencyType = dependencyType;
        }

        public Task OnRegistrationAsync(IModuleRegistrationContext context)
        {
            context.AddDependency(_dependencyType);
            return Task.CompletedTask;
        }
    }

    public class ModuleA : Module<string>
    {
        protected internal override async Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            ExecutionOrder.Add("A");
            await Task.Yield();
            return "A";
        }
    }

    [AddDependency(typeof(ModuleA))]
    public class ModuleB : Module<string>
    {
        protected internal override async Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            ExecutionOrder.Add("B");
            await Task.Yield();
            return "B";
        }
    }

    // Never registered — a dynamic dependency on this module cannot be satisfied.
    public class UnregisteredModule : Module<string>
    {
        protected internal override async Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "unregistered";
        }
    }

    // The dependency is added at registration time (after build-time auto-registration has run),
    // so nothing auto-registers UnregisteredModule and it must be caught by the run-time revalidation.
    [AddDependency(typeof(UnregisteredModule))]
    public class ModuleWithMissingDynamicDependency : Module<string>
    {
        protected internal override async Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "never runs";
        }
    }

    [ModularPipelines.Attributes.ModuleCategory("compile")]
    public class DynamicallySkippedDependency : Module<string>
    {
        protected internal override Task<string> ExecuteAsync(
            IModuleContext context,
            CancellationToken cancellationToken) =>
            throw new InvalidOperationException("A filtered dependency must not execute");
    }

    [ModularPipelines.Attributes.ModuleCategory("test")]
    [AddDependency(typeof(DynamicallySkippedDependency))]
    public class DynamicallySkippedDependent : Module<string>
    {
        protected internal override Task<string> ExecuteAsync(
            IModuleContext context,
            CancellationToken cancellationToken) =>
            throw new InvalidOperationException("A cascade-skipped dependent must not execute");
    }

    [Before(Test)]
    public void ClearExecutionOrder()
    {
        ExecutionOrder.Clear();
    }

    [Test]
    public async Task DynamicDependency_ModuleBWaitsForModuleA()
    {
        var result = await TestPipelineBuilder.Create()
            .AddModule<ModuleA>()
            .AddModule<ModuleB>()
            .RunAsync();

        await Assert.That(result.Status).IsEqualTo(Enums.ModuleStatus.Succeeded);
        await Assert.That(ExecutionOrder).IsEquivalentTo(new[] { "A", "B" });
    }

    [Test]
    public async Task DynamicDependency_OnUnregisteredModule_FailsFast()
    {
        // A dependency added via a registration event on a module that is not registered
        // (and cannot be auto-registered, because it is not declared via [DependsOn]) must be
        // caught by the run-time revalidation of the canonical graph, before the scheduler runs,
        // rather than surfacing late as a dependency-waiter failure.
        await Assert.ThrowsAsync<ModuleNotRegisteredException>(() =>
            TestPipelineBuilder.Create()
                .AddModule<ModuleWithMissingDynamicDependency>()
                .RunAsync());
    }

    [Test]
    public async Task DynamicDependency_OnFilteredModule_CascadeSkipsDependent()
    {
        var result = await TestPipelineBuilder.Create()
            .AddModule<DynamicallySkippedDependency>()
            .AddModule<DynamicallySkippedDependent>()
            .ConfigurePipelineOptions(options => options with { RunOnlyCategories = ["test"] })
            .RunAsync();

        await Assert.That(result.Status).IsEqualTo(Enums.ModuleStatus.Succeeded);

        var dependentResult = await result.Modules
            .OfType<DynamicallySkippedDependent>()
            .Single();
        await Assert.That(dependentResult.SkipDecisionOrDefault).IsNotNull();
        await Assert.That(dependentResult.SkipDecisionOrDefault!.Reason)
            .Contains(nameof(DynamicallySkippedDependency));
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/DynamicDependencyIntegrationTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open System
open System.Threading
open System.Threading.Tasks
open ModularPipelines.Attributes.Events
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module DynamicDependencyIntegrationTests =
    let executionOrder = ResizeArray<string>()

    type AddDependencyAttribute(dependencyType: Type) =
        inherit Attribute()
        interface IModuleRegistrationEventReceiver with
            member _.OnRegistrationAsync(context: IModuleRegistrationContext) =
                context.AddDependency(dependencyType)
                Task.CompletedTask

    type ModuleA() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                executionOrder.Add("A")
                do! Task.Yield()
                return "A"
            }

    [<AddDependency(typeof<ModuleA>)>]
    type ModuleB() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                executionOrder.Add("B")
                do! Task.Yield()
                return "B"
            }

[<NotInParallel(nameof DynamicDependencyIntegrationTests)>]
type DynamicDependencyIntegrationTests() =
    inherit TestBase()

    [<Before(HookType.Test)>]
    member _.ClearExecutionOrder() =
        DynamicDependencyIntegrationTests.executionOrder.Clear()

    [<Test>]
    member _.DynamicDependency_ModuleBWaitsForModuleA() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyIntegrationTests.ModuleA>()
                .AddModule<DynamicDependencyIntegrationTests.ModuleB>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())
        do! check(Assert.That((DynamicDependencyIntegrationTests.executionOrder |> Seq.toArray) = [| "A"; "B" |]).IsTrue())
    }
